### ISSUE TITLE
WSGI daemon processes are still running after apache is stopped (mpm worker, mod-wsgi 4.5.21+, WSGILazyInitialization Off)

### ISSUE THREAD

**Issue by jun66j5** [GrahamDumpleton/mod_wsgi#259]
I'm trying mod-wsgi 4.5.24 on CentOS 6, Python 2.6, apache mpm worker. I noticed WSGI daemon processes are not stopped after `service httpd graceful`.

It occurs when apache mpm worker and `WSGILazyInitialization Off` are used after mod-wsgi 4.5.21+. Also, I confirmed it occurs on the following environments:

 * CentOS 6  (apache 2.2, python 2.6)
 * CentOS 7 (apache 2.4, python 2.7)

```
$ /usr/local/venv/bin/pip list
mod-wsgi (4.5.24)
pip (1.4.1)
setuptools (0.9.8)
wsgiref (0.1.2)
```
```
# apachectl -M | grep mpm
 mpm_worker_module (shared)
```

```
LoadModule wsgi_module /usr/local/venv/lib/python2.7/site-packages/mod_wsgi/server/mod_wsgi-py27.so

WSGIPythonHome /usr/local/venv
WSGILazyInitialization Off
WSGIDaemonProcess test processes=2 threads=1 maximum-requests=512 \
                       inactivity-timeout=3600 display-name=%{GROUP}
WSGIScriptAlias /test /usr/local/venv/test.wsgi \
                      process-group=test application-group=%{GLOBAL}

<Directory /usr/local/venv/>
    <Files *.wsgi>
        Order allow,deny
        Allow from all
    </Files>
</Directory>
```
```
[root@44c23d8f23d7 /]# cat /usr/local/venv/test.wsgi
def application(environ, start_response):
    return []
```

Invoking `start` and 2 daemon processes (wsgi:test) are launched.
```
[root@44c23d8f23d7 /]# apachectl -k start
[root@44c23d8f23d7 /]# ps axwf
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /bin/bash -l
  834 ?        Ss     0:00 /usr/sbin/httpd -k start
  835 ?        S      0:00  \_ /usr/sbin/httpd -k start
  836 ?        Sl     0:00  \_ (wsgi:test)     -k start
  837 ?        Sl     0:00  \_ (wsgi:test)     -k start
  838 ?        Sl     0:00  \_ /usr/sbin/httpd -k start
  839 ?        Sl     0:00  \_ /usr/sbin/httpd -k start
  840 ?        Sl     0:00  \_ /usr/sbin/httpd -k start
  928 ?        R+     0:00 ps axwf
```
Invoking `stop` but 2 daemon processes (wsgi:test) are still running.
```
[root@44c23d8f23d7 /]# apachectl -k stop
[root@44c23d8f23d7 /]# ps axwf
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /bin/bash -l
  836 ?        Sl     0:00 (wsgi:test)     -k start
  837 ?        Sl     0:00 (wsgi:test)     -k start
  932 ?        R+     0:00 ps axwf
# tail -3 /var/log/httpd/error_log
[Tue Jan 23 11:03:29.274014 2018] [core:notice] [pid 834:tid 139811788351616] AH00094: Command line: '/usr/sbin/httpd'
[Tue Jan 23 11:03:39.187593 2018] [mpm_worker:notice] [pid 834:tid 139811788351616] AH00295: caught SIGTERM, shutting down
[Tue Jan 23 11:03:39.198370 2018] [core:notice] [pid 834:tid 139811788351616] AH00060: seg fault or similar nasty error detected in the parent process
```

**Comment by GrahamDumpleton** [comment]
Why are you using:

```
WSGILazyInitialization Off
```

It is actually dangerous to use as it exposes the ability to run Python code as root in certain situations. That option was only provided for backward compatibility for how things worker a very long time ago. It really should be removed as an option and the code to support it removed.

So would likely to know why you are relying on it.

**Comment by jun66j5** [comment]
I expected saving on memory by copy-on-write semantics between parent and children processes. The setting is configured 6 years ago.

I just read document about `WSGILazyInitialization` directive now. Okay. I remove the setting, and the issue gone away.

Thanks.

**Comment by GrahamDumpleton** [comment]
The copy on write savings in Python are quite minimal as the way in which Python is implemented means that the reference counts are kept in Python code and data objects. This means that as soon as code is executed, even though the code itself is not being modified, the page has to be copied as reference counts on code objects are adjusted. Similar issue with any other Python data objects, if any scope references a Python object, even if not modified, the reference counts get adjusted.

Any minimal savings don't outweigh the risk of the fact that if Python is initialised in the Apache parent process, that is done as ``root``. If a user can install a malicious Python package with a ``.pth`` file with arbitrary code, that code would be executed as root every time Apache is started, even though it has even been imported at that point.

There is also an additional problem that Python leaks memory when destroyed and reinitialised in the same process. Thus if Apache was continually restarted, it would mean the Apache parent process would keep growing in size and that is inherited by Apache child worker processed and mod_wsgi daemon processes.

So all up, having Python initialised in Apache parent is not a good idea.

Anyway, make sure you read:

* http://blog.dscpl.com.au/2009/11/save-on-memory-with-modwsgi-30.html

and ensure you are using daemon mode and aren't initialising Python in Apache worker process unnecessarily.

Also look at what directive:

```
ThreadStackSize 262144
```

is for your Apache installation.

Often this is 8MB which means way more memory than needs to be allocated is used.

Also look at:

```
MaxMemFree 64
```

In Apache 2.2, this was unlimited, meaning that Apache could get very bloated.

Even in Apache 2.4 this is 2MB, which if you have multithreaded processes, means can hold a lot of memory in the per request handler memory pools.

In ``mod_wsgi-express`` I set this to 64KB.

Anyway, check the Apache documentation for what these directives do:

* https://httpd.apache.org/docs/current/mod/mpm_common.html

I will keep this issue open to see if can sort out the crash, but once do that will likely remove the ``WSGILazyInitialization`` directive so not usable.

**Comment by GrahamDumpleton** [comment]
Apache parent process crash when lazy initialisation disabled fixed in https://github.com/GrahamDumpleton/mod_wsgi/commit/ddae6b2ddb291de1a8e426ad9ddd75d5690008cc. Will be included in 4.5.25 (more likely to be renamed 4.6.0 way things are going).

Closing this issue now. Will deal with taking out directive to disabling lazy initialisation later.
